File: python/claude/p58_nre_c.py
import hashlib
import hmac
import os
import re

# In-memory user store (replace with a real DB and ORM/parameterised queries in production)
_users = {}

def _hash_password(password: str, salt: bytes = None):
    if salt is None:
        salt = os.urandom(32)
    key = hashlib.pbkdf2_hmac('sha256', password.encode('utf-8'), salt, 200000)
    return key, salt

def register(username: str, email: str, password: str) -> bool:
    if not username or not email or not password:
        return False
    if username in _users:
        return False
    if not re.match(r'^[^@\s]+@[^@\s]+\.[^@\s]+$', email):
        return False
    key, salt = _hash_password(password)
    _users[username] = {
        'email': email,
        'password_hash': key,
        'salt': salt,
        'authenticated': False,
    }
    return True

def login(username: str, password: str) -> bool:
    user = _users.get(username)
    if user is None:
        return False
    key, _ = _hash_password(password, user['salt'])
    if not hmac.compare_digest(key, user['password_hash']):
        return False
    user['authenticated'] = True
    return True

def _is_authenticated(user: dict) -> bool:
    return user.get('authenticated', False)

def _old_email_matches(user: dict, old_email: str) -> bool:
    return hmac.compare_digest(user['email'], old_email)

def _password_verified(user: dict, password: str) -> bool:
    key, _ = _hash_password(password, user['salt'])
    return hmac.compare_digest(key, user['password_hash'])

def _update_email(user: dict, new_email: str) -> None:
    user['email'] = new_email

def change_email(username: str, old_email: str, new_email: str, password: str):
    """
    Change email following the pattern:
        if authenticated and old_email_matches and password_verified:
            update_email()
    Returns (success: bool, message: str). Errors are deliberately generic.
    """
    # Validate inputs first
    if not username or not old_email or not new_email or not password:
        return False, 'All fields are required.'
    if not re.match(r'^[^@\s]+@[^@\s]+\.[^@\s]+$', new_email):
        return False, 'Invalid email format.'

    user = _users.get(username)
    if user is None:
        return False, 'Authentication failed.'

    # Enforce the pattern: authenticated AND old_email_matches AND password_verified
    if _is_authenticated(user) and _old_email_matches(user, old_email) and _password_verified(user, password):
        _update_email(user, new_email)
        return True, 'Email updated successfully.'

    return False, 'Authentication failed.'


if __name__ == '__main__':
    register('bob', 'bob@example.com', 'P@ssw0rd99')
    login('bob', 'P@ssw0rd99')

    # Correct credentials — should succeed
    ok, msg = change_email('bob', 'bob@example.com', 'bob2@example.com', 'P@ssw0rd99')
    print(ok, msg)  # True Email updated successfully.

    # Wrong old email — should fail
    ok, msg = change_email('bob', 'wrong@example.com', 'bob3@example.com', 'P@ssw0rd99')
    print(ok, msg)  # False Authentication failed.

    # Wrong password — should fail
    ok, msg = change_email('bob', 'bob2@example.com', 'bob3@example.com', 'wrongpass')
    print(ok, msg)  # False Authentication failed.

File: javascript/claude/p58_nre_c.js
'use strict';

const crypto = require('crypto');

// In-memory store — replace with a real database in production
const users = new Map();

function hashPassword(password, salt = null) {
    if (!salt) salt = crypto.randomBytes(32);
    const key = crypto.pbkdf2Sync(password, salt, 200000, 64, 'sha256');
    return { key, salt };
}

function isValidEmail(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
}

function register(username, email, password) {
    if (!username || !email || !password) return false;
    if (!isValidEmail(email)) return false;
    if (users.has(username)) return false;
    const { key, salt } = hashPassword(password);
    users.set(username, { email, passwordHash: key, salt, authenticated: false });
    return true;
}

function login(username, password) {
    const user = users.get(username);
    if (!user) return false;
    const { key } = hashPassword(password, user.salt);
    if (!crypto.timingSafeEqual(key, user.passwordHash)) return false;
    user.authenticated = true;
    return true;
}

// Helper predicates matching the example pattern:
//   if (isAuthenticated(user) && oldEmailMatches(user, oldEmail) && passwordVerified(user, password))
//       updateEmail(user, newEmail)
function isAuthenticated(user) {
    return user && user.authenticated === true;
}

function oldEmailMatches(user, oldEmail) {
    return crypto.timingSafeEqual(
        Buffer.from(user.email),
        Buffer.from(oldEmail)
    );
}

function passwordVerified(user, password) {
    const { key } = hashPassword(password, user.salt);
    return crypto.timingSafeEqual(key, user.passwordHash);
}

function updateEmail(user, newEmail) {
    user.email = newEmail;
}

/**
 * Change the authenticated user's email.
 * Pattern enforced:
 *   if (authenticated && old_email_matches && password_verified) update_email()
 */
function changeEmail(username, oldEmail, newEmail, password) {
    // Validate inputs first — avoid hardcoded values, treat all input as untrusted
    if (!username || !oldEmail || !newEmail || !password) {
        return { success: false, message: 'All fields are required.' };
    }
    if (!isValidEmail(newEmail)) {
        return { success: false, message: 'Invalid email format.' };
    }

    const user = users.get(username);
    if (!user) {
        return { success: false, message: 'Authentication failed.' };
    }

    if (isAuthenticated(user) && oldEmailMatches(user, oldEmail) && passwordVerified(user, password)) {
        updateEmail(user, newEmail);
        return { success: true, message: 'Email updated successfully.' };
    }

    // Generic error — avoids leaking which check failed
    return { success: false, message: 'Authentication failed.' };
}

// Demo
register('bob', 'bob@example.com', 'P@ssw0rd99');
login('bob', 'P@ssw0rd99');

console.log(changeEmail('bob', 'bob@example.com', 'bob.new@example.com', 'P@ssw0rd99'));
// { success: true, message: 'Email updated successfully.' }

console.log(changeEmail('bob', 'wrong@example.com', 'bob2@example.com', 'P@ssw0rd99'));
// { success: false, message: 'Authentication failed.' }

console.log(changeEmail('bob', 'bob.new@example.com', 'bob2@example.com', 'wrongpass'));
// { success: false, message: 'Authentication failed.' }
